src/user/tag.ts

```diff
@@ -0,0 +1,202 @@
+      )
+    }
+
+    if (!this.puppet) {
+      throw new Error('Tag class can not be instanciated without a puppet!')
+    }
+  }
+
+  /**
+   * @private
+   * About the Generic: https://stackoverflow.com/q/43003970/1123955
+   *
+   * Get Tag by id
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @param {string} id
+   * @returns {Tag}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * const tag = bot.Tag.load('tagId')
+   */
+  public static load<T extends typeof Tag> (
+    this : T,
+    id   : string,
+  ): T['prototype'] {
+    if (!this.pool) {
+      log.verbose('Tag', 'load(%s) init pool', id)
+      this.pool = new Map<string, Tag>()
+    }
+    if (this === Tag) {
+      throw new Error(
+        'The global Tag class can not be used directly!'
+        + 'See: https://github.com/Chatie/wechaty/issues/1217',
+      )
+    }
+    if (this.pool === Tag.pool) {
+      throw new Error('the current pool is equal to the global pool error!')
+    }
+    const existingTag = this.pool.get(id)
+    if (existingTag) {
+      return existingTag
+    }
+
+    const newTag = new (this as any)(id) as Tag
+
+    this.pool.set(id, newTag)
+
+    return newTag
+  }
+
+  /**
+   * Get all tags of this account
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @returns {Promise<Tag[]>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * await bot.Tag.findAll()
+   */
+  public static async findAll<T extends typeof Tag> (this: T): Promise<Array<T['prototype']>> {
+    log.verbose('Tag', 'findAll()')
+
+    try {
+      const tags = await this.puppet.allTags()
+      const invalidDict: { [id: string]: true } = {}
+
+      const tagList = tags.map(tag => {
+        const newTag = this.load(tag.id)
+        newTag.payload = tag
+        return newTag
+      })
+      return tagList.filter(tag => !invalidDict[tag.id])
+    } catch (e) {
+      log.error('Tag', 'findAll() exception: %s', e.message)
+      throw new Error(`findAll error : ${e}`)
+    }
+  }
+
+  /**
+   * Get a new tag
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @param {string} [tag] the tag name which want to create
+   * @returns {Promise<Tag>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * await bot.Tag.get(tag)
+   */
+  public static async get<T extends typeof Tag> (this: T, tag: string): Promise<T['prototype']> {
+    log.verbose('Tag', 'get()')
+
+    try {
+      const tagPayload: TagPayload = await this.puppet.createTag(tag)
```

Comment:
If the tag exists, we should not create a new tag here, right?

Reply:
It will return the existed tag instead of creating.

Reply:
Then the function is misleading. We shouldn't call it `createTag`, it should be `getOrCreateTag` or something else. `createTag` explicitly tell the user that this is creating a tag.

Reply:
Agree, `getOrCreateTag()` is better for the puppet method.

Reply:
I guess @huan might not like this design :P

Let's wait for his review.